include/array/array.h

```diff
@@ -2264,12 +2264,32 @@ class array {
    * `alloc`. This is a deep copy of the contents of `other`. */
   array(const array& other, const Alloc& alloc) : array(alloc) { assign(other); }
 
+  /** Move construct from another array `other`. If the allocator of this array
+   * is propagate on container move assignment, this operation moves the allocation
+   * of other to this array, and the other array becomes a default constructed
+   * array. If the allocator of this and the other array are non-equal, each
+   * element is move-constructed into a new allocation. */
+  array(array&& other)
+      : alloc_(Alloc()), buffer_(nullptr), buffer_size_(0),
```

Comment:
Isn't this still default constructing an allocator even when it isn't necessary? I think issue #47 still exists with this fix.

Reply:
I just didn't realize we can get away with not initializing `alloc_`. Initializing it or not has no effect on `propagate_on_container_move_assignment::value`. Removed.

I focused on the copy part of the move constructor. With the latest code, #47 should be fixed.

Reply:
Since `Alloc alloc_` is a member variable of the class array, I guess the allocator will be default constructed no matter what?

Maybe we can just remove the claim that it fixes #47, and get this in?

Reply:
Hmm, I think you might be right that this actually does fix #47. Before, the extra allocator was constructed in order to call the other constructor, and then move/copy construct the member variable from that (I guess). With this change, that might not happen. The member allocator will be constructed either way.

However, if it doesn't, is this change an improvement in some other way?

I think we need to write a test that answers this question (and prevents the reoccurrence of #47 in the future).

Reply:
> However, if it doesn't, is this change an improvement in some other way?

The move constructor no longer copies the array by default when using a custom allocator.

Hmm, the array_lifetime test is disabled and broken. Let's see if I can fix it...

Reply:
To be clear,
```
array src(shape, my_custom_allocator());
// This shouldn't copy the array but it does
array dst = std::move(src);
// This doesn't copy the array, but it's counter-intuitive and not always possible.
array dst2 = array(std::move(src), my_custom_allocator());
```

Reply:
Oh, I see, the current array move constructor uses (in)equality of the allocators to decide if the values should be copied or moved. In your change, it uses `propagate_on_container_move_assignment::value` to decide to do this instead. The whole reason #47 was hard to fix was because allocator equality couldn't be determined at compile time.

I remember spending a lot of time on this kind of issue, so I'd want to understand for sure if this is the right behavior.

It looks to me like the array_lifetime tests are enabled and working? I just ran them on master, and it looks OK.

Reply:
I think I got the behavior I'm using now (allocator equality) from emulating `std::vector` as described here: https://cplusplus.com/reference/vector/vector/vector/:

> If alloc is specified and is different from x's allocator, the elements are moved. Otherwise, no elements are constructed (their ownership is directly transferred).

Of course, that description is a bit vague...

Reply:
> I remember spending a lot of time on this kind of issue, so I'd want to understand for sure if this is the right behavior.

Sure let's get this right.

> How did you run the test? array_lifetime is missing from this test list.

https://github.com/dsharlet/array/blob/master/test/BUILD.bazel#L7

Reply:
Oops, that's the bazel build (@jiawen). In the make build, it's enabled and working (I just tested it): https://github.com/dsharlet/array/blob/master/Makefile#L7

Reply:
Just sent https://github.com/dsharlet/array/pull/89 to address that issue, thanks for pointing it out!

Reply:
Np.

Oops I see. My changes broke the array_lifetime test. The auto_allocator is tricky. Let me fix it.

> I think I got the behavior I'm using now (allocator equality) from emulating std::vector as described here: https://cplusplus.com/reference/vector/vector/vector/:

I think the spec can be interpreted as when alloc is not specified, no elements are constructed.

Reply:
Reading the libstdc++ implementation of `std::vector`: it just straight up doesn't care about the possibility of an allocator not supporting moves, and just always swaps :(. I think this is the problem I ran into here, and the logic I have currently was my best attempt at working around the lack of proper support for this in (at least one common) STL implementation.

Reply:
Alright, PTAL. Unit test fixed. The newly added test (test_move_lifetime with stateful_movable_custom_alloc) fails with the old move constructor impl.

The compiler is complaining about the use of constexpr before C++17. I used it in the test regardless to avoid code duplication. If C++17 is the official requirement, we can add constexpr to a lot of the if's in array.h.